---
title: "Statistical Analysis - Loa Proteome"
authors: "Clemens Dierks"
orcid: 0000-0002-4560-8939
email: clemens.dierks@charite.de 
date: last-modified 
format: 
    html:
        code-fold: true
        code-overflow: wrap
        code-tools: true
        graphics: yes
        toc: true
        toc-location: left-body
        toc-title: "Outline"
editor: 
  markdown: 
    wrap: 72
---

## Overview

In this script, we conduct differential expression analysis to compare
loa- infected patients with control patients, as well as to evaluate
differences between various infection stages. These stages include
microfilarial infection only (MF), eye worm + microfilarial infection
(EWMF), eye worm infection only (EW). Additionally, we analyze the
statistical differences in symptom prevalence between controls and
EW-infected patients. We also perform a trend analysis across distinct
groups: control asymptomatic, control symptomatic, EW asymptomatic, and
EW symptomatic.

Subsequent analysis will be carried out in Python.

## Imports

```{r}
library(limma)
library(stringr)
library(MatchIt)
library(edgeR)
library(org.Hs.eg.db)
library(DOSE) # disease ontologyn
library(clinfun)
library(dplyr)
library(openxlsx)
library(writexl)

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir =  "C:/Users/cleme/OneDrive - Charité - Universitätsmedizin Berlin/Projects/02_NTD_CHARITE/NTD")
```



## Data Import

```{r}
# load data
preprocessed_file_path <- "data/analysis/loa/loa_data.csv"
data <- read.table(preprocessed_file_path, sep=",", header = T, check.names = F)
imputed_file_path <- "data/analysis/loa/loa_data_imp.csv"
data_imp <- read.table(imputed_file_path, sep=",", header = T, check.names = F)

# get protein names
proteins <- read.table("data/preprocessing_boris/NTD_min1peptides/NTD_min1peptides_ProteinMaxLFQ_reprep_logtransformed.tsv", sep="\t", check.names=FALSE)
colnames(proteins) <- proteins[3, ]
row.names(proteins) <- NULL
protein_map <- proteins[4:length(row.names(proteins)),1:2]

# common uniprot ids
common_uniprot <- colnames(data)[99:345]
common_genenames <- protein_map %>% filter(Protein.Group %in% common_uniprot)
common_genenames <- common_genenames$Genes

# Convert all protein columns to numeric
data[, common_uniprot] <- as.data.frame(sapply(data[, common_uniprot], as.numeric))

data <- data[(data$infection=="loa") | (data$infection=="healthy controls"), ]

data <- data[which((data$eos_abs < 0.5 & data$`disease severity` == "healthy control") | data$`disease severity` != "healthy control"), ]

data$disease_severity <- data$`disease severity`
data[data$disease_severity == "healthy control", "disease_severity"] <- "CTRL"
data[data$disease_severity == "HYMF", "disease_severity"] <- "MF"

data[data$infection == "healthy controls", "infection"] <- "CTRL"
```

## Differential Expression Analysis

```{r}
design_infection <- model.matrix(~0+infection, data=data)
colnames(design_infection) <- gsub("infection", "", colnames(design_infection))
design_infection <- design_infection[,c("CTRL", "loa")]


common_uniprot <- colnames(data)[99:345]

data[,colnames(data) %in% common_uniprot] <- as.data.frame(sapply(data[colnames(data) %in% common_uniprot], as.numeric))
fit_infection <- lmFit(t(data[,colnames(data) %in% common_uniprot]), design = design_infection)
colnames(fit_infection$coefficients)

cont.matrix_infection <- makeContrasts(
  loa-CTRL,
  levels=design_infection)

fit2_infection <- contrasts.fit(fit_infection, cont.matrix_infection)
fit3_infection <- eBayes(fit2_infection, trend = T)

res_infection <- topTable(fit3_infection, n=Inf, adjust="BH")

rownames(protein_map) <- protein_map$Protein.Group
res_infection$genes <- protein_map[row.names(res_infection), "Genes"]

write.csv(res_infection, "results/loaloa/tables/dea_limma/ctrl_infected.csv")

plot(res_infection$logFC, -log10(res_infection$P.Value))
plot(res_infection$logFC, -log10(res_infection$adj.P.Val))
```

```{r}
# Perform Clusterprofiler analysis 

uniprot_ids <- rownames(res_infection[res_infection$adj.P.Val<0.05,])
entrez_ids <- bitr(rownames(res_infection[res_infection$adj.P.Val<0.05,]), fromType = "UNIPROT", toType = "ENTREZID", OrgDb = org.Hs.eg.db)$ENTREZID

ego <- enrichGO( gene = rownames(res_infection[res_infection$adj.P.Val<0.05,]),
          OrgDb = org.Hs.eg.db,
          keyType = "UNIPROT",
          ont = "BP",
          pAdjustMethod = "none",
          )
edo <- enrichDO(gene         = entrez_ids,
                ont          = "DO",
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.05)


dotplot(ego)
dotplot(edo)
```

############ 

```{r}
### Severity Groups

data$disease_severity <- as.factor(data$disease_severity)
design <- model.matrix(~0+disease_severity, data=data)
colnames(design) <- sub("disease_severity", "", colnames(design))
colnames(design) <- sub(":", "_", colnames(design))

#change uniprot ids to gene names
# drop columns with non mappable column name
# data <- data[, !colnames(data) == "", drop=FALSE]
data[,colnames(data) %in% common_uniprot] <- as.data.frame(sapply(data[,colnames(data) %in% common_uniprot], as.numeric))
fit <- lmFit(t(data[,colnames(data) %in% common_uniprot]), design = design)

table(data$disease_severity)


# Define contrasts
cont.matrix <- makeContrasts(
  MF-EW,
  MF-EWMF,
  EW-EWMF,
  levels=design)


fit2 <- contrasts.fit(fit, cont.matrix)
fit3 <- eBayes(fit2, trend = T, robust = T)

res_mf_ew <- topTable(fit3, n=Inf, adjust="BH", coef = 1)
res_mf_ewmf <- topTable(fit3, n=Inf, adjust="BH", coef = 2)
res_ew_ewmf <- topTable(fit3, n=Inf, adjust="BH", coef = 3)

res_mf_ew$genes <- protein_map[row.names(res_mf_ew), "Genes"]
res_mf_ewmf$genes <- protein_map[row.names(res_mf_ewmf), "Genes"]
res_ew_ewmf$genes <- protein_map[row.names(res_ew_ewmf), "Genes"]

write.csv(res_mf_ew, "results/loaloa/tables/dea_limma/mf_ew.csv")
write.csv(res_mf_ewmf, "results/loaloa/tables/dea_limma/mf_ewmf.csv")
write.csv(res_ew_ewmf, "results/loaloa/tables/dea_limma/ew_ewmf.csv")


```

```{r}
## only with data with raploa 1 year
data_rl <- data[((data$disease_severity == "EW") & (data$rl_lastyear == 1)) | ((data$disease_severity == "EWMF") & (data$rl_lastyear == 1)) | (data$disease_severity == "MF"),]

design <- model.matrix(~0+disease_severity, data=data_rl)
colnames(design) <- sub("disease_severity", "", colnames(design))
fit <- lmFit(t(data_rl[,colnames(data_rl) %in% common_uniprot]), design = design)


table(data_rl$disease_severity)

# Define contrasts
cont.matrix <- makeContrasts(
  MF-EW,
  MF-EWMF,
  EW-EWMF,
  levels=design)


fit2 <- contrasts.fit(fit, cont.matrix)
fit3 <- eBayes(fit2, trend = T, robust = T)

res_mf_ew_rl <- topTable(fit3, n=Inf, adjust="BH", coef = 1)
res_mf_ewmf_rl <- topTable(fit3, n=Inf, adjust="BH", coef = 2)
res_ew_ewmf_rl <- topTable(fit3, n=Inf, adjust="BH", coef = 3)

res_mf_ew_rl$genes <- protein_map[row.names(res_mf_ew_rl), "Genes"]
res_mf_ewmf_rl$genes <- protein_map[row.names(res_mf_ewmf_rl), "Genes"]
res_ew_ewmf_rl$genes <- protein_map[row.names(res_ew_ewmf_rl), "Genes"]

write.csv(res_mf_ew_rl, "results/loaloa/tables/dea_limma/mf_ew_rl.csv")
write.csv(res_mf_ewmf_rl, "results/loaloa/tables/dea_limma/mf_ewmf_rl.csv")
write.csv(res_ew_ewmf_rl, "results/loaloa/tables/dea_limma/ew_ewmf_rl.csv")

plot(res_mf_ew_rl$logFC, -log10(res_mf_ew_rl$P.Value))
plot(res_mf_ewmf_rl$logFC, -log10(res_mf_ew_rl$P.Value))
plot(res_ew_ewmf_rl$logFC, -log10(res_mf_ew_rl$P.Value))

plot(res_mf_ew_rl$logFC, -log10(res_mf_ew_rl$adj.P.Val))
plot(res_mf_ewmf_rl$logFC, -log10(res_mf_ewmf_rl$adj.P.Val))
plot(res_ew_ewmf_rl$logFC, -log10(res_ew_ewmf_rl$adj.P.Val))

```

## Trend Analysis of control and Ew positive patients

```{r}
dim(data[data["pru_history_severity"]=="Ctrl asympt.",])[1]
```


```{r}

data <- as.data.frame(data)

clinical_symptoms_bin <- colnames(data)[382:411]

df_res_list <- list()

for (i in 1:length(clinical_symptoms_bin)) {
  c <- clinical_symptoms_bin[i]
  print(c)
  
    
  data_symp <- data[
    (data[c] == "Ctrl asympt.") | (data[c] == "Ctrl sympt.") | (data[c] == "EW asympt.") | (data[c] == "EW sympt."),]
  data_symp[data_symp[c] == "Ctrl asympt.", c] = 1
  data_symp[data_symp[c] == "Ctrl sympt.", c] = 2
  data_symp[data_symp[c] == "EW asympt.", c] = 3
  data_symp[data_symp[c] == "EW sympt.", c] = 4
  
  # Functions to conduct kruskal wallis and jonkheere - terpstra test
  kw_protein <- function(p_col, g_col) {
    # vec <- split(data_symp[[p_col]], data_symp[[g_col]])
    # print(vec)
    df <- data_symp[complete.cases(data_symp[, p_col]),]
    test_res <- kruskal.test(x=df[,p_col], g=as.numeric(df[,g_col]))
    
    # test_res <- kruskal.test(vec)
    return(
      list(
        KW_statistic = test_res$statistic,
        KW_p_value = test_res$p.value,
        n_ctrl_asympt = dim(df[df[g_col]==1,])[1],
        n_ctrl_sympt = dim(df[df[g_col]==2,])[1],
        n_ew_asympt = dim(df[df[g_col]==3,])[1],
        n_ew_sympt = dim(df[df[g_col]==4,])[1],
        median_ctrl_asympt = median(df[df[g_col]==1, p_col]),
        median_ctrl_sympt = median(df[df[g_col]==2, p_col]),
        median_ew_asympt = median(df[df[g_col]==3, p_col]),
        median_ew_sympt = median(df[df[g_col]==4, p_col])
      )
    )
  }
  jt_protein <- function(p_col, g_col, order) {
    df <- data_symp[complete.cases(data_symp[, p_col]),]
    test_res <- jonckheere.test(x=df[,p_col], g=as.numeric(df[,g_col]), nperm = 1000)
    return(
      list(
        JT_statistic = test_res$statistic,
        JT_p_value = test_res$p.value
      )
    )
  }
  
  # apply tests to each protein
  results_list_kruskal <- lapply(common_uniprot, function (p_col) {
    kw_protein(p_col, c)
  })
  
  
  results_list_jonkheere <- lapply(common_uniprot, function (p_col) {
    jt_protein(p_col, c)
  })
  
  results_df_kruskal <- do.call(
    rbind, lapply(results_list_kruskal, as.data.frame))
  results_df_kruskal$KW_adj_pvalue <- p.adjust(
    results_df_kruskal$KW_p_value, method = "BH")
  
  results_df_jonkheere <- do.call(
    rbind, lapply(results_list_jonkheere, as.data.frame))
  results_df_jonkheere$JT_adj_pvalue <- p.adjust(
    results_df_jonkheere$JT_p_value, method = "BH")
  
  res_kw_jt <- cbind(results_df_kruskal, results_df_jonkheere)
  res_kw_jt$Uniprot <- common_uniprot
  res_kw_jt$Genes <- common_genenames
  
  # collect results in list of dataframes
  c_temp <- gsub("_severity", "", clinical_symptoms_bin[i])
  df_res_list[[c_temp]] <- res_kw_jt

}
# Export  
write.xlsx(df_res_list, "results/loaloa/tables/sympt_stat_analysis.xlsx")
```
